### PR TITLE
Add the bot cleanup worker

### LINKED ISSUE #3: We should cleanup the messages in the chatroom when they are no longer needed
The bot is spammy. We should delete messages after being posted.

Deleting messages feel like a good next move to address the bot dominating the channel.

Based on the official documentation: "A message can only be deleted if it was sent less than 48 hours ago".
Based on the suggested usage, I think this limitation will allow us to remove a good chunk of the messages after verification

Since captchas expire much faster we should be fine.

### PR DESCRIPTION
**Pull request by hhio618**
Fixes #3 
Add a cleanup worker that runs every 45 hours (<48 hours of the telegram limit) and delete sent messages from the bot.
I think I need to add some tests, then we should be good to go.

### CODE REVIEW

**Comment by posix4e** [comment]
How you test it on telegram?

**Comment by hhio618** [comment]
Hi @posix4e, You can use the following steps:
1- Setup a telegram bot and an hCaptcha account.
1- Create a Postgres DB instance and migrate.
2- Start flask application -> the bot and the worker will be started in separate threads.
3- Every sent message will be recorded (chat_id, message_id).
4- The cleanup worker runs every 45 hours and cleans the old bot messages (This can be changed to seconds for testing purposes).

Also, I added some test cases too.

**Comment by gaieges** [comment]
I'm curious why we don't just delete after the siteverify has been performed and the user has been activated in the telegram channel?  Is there another async step that I'm missing here that is out of our control?

**Comment by hhio618** [comment]
Of course, we can just delete the message right after the user has been activated. but there might be a chance the user couldn't see the verification message. Is that ok?
So do we need to Immediately delete the verification message and the corresponding siteverify message?
Can you please specify which messages should be deleted Immediately?

**Comment by gaieges** [comment]
I think we mostly just care about the messages in the channel, correct @posix4e ?  

Messages to the user directly can probably stay around forever, I'd think.

**Comment by posix4e** [comment]
> I think we mostly just care about the messages in the channel, correct @posix4e ?
> 
> Messages to the user directly can probably stay around forever, I'd think.

True

**Comment by posix4e** [comment]
Let's change the pr to reflect that. We wanna delete messages after two minutes or if it's solved

**Review by posix4e — DISMISSED** [review]
The behavior.is to delete the messages as they are solved/failed or expired (2 minutes) 45 hours is too long to wait

**Comment by gaieges** [comment]
Another note around this: the cronjob functionality is a good idea that we can use down the road, so it's good to keep that with the above adjustments

**Comment by posix4e** [comment]
Yea perhaps for a different pr though .

>

**Comment by gaieges** [comment]
@hhio618 there's a bit of functionality that already sort of achieves a "cronjob" like approach here, so all we're saying is it's probably fine the way it is, with the adjustments of:

- Just track channel messages and clean em up after successful siteverify
- If no followup or captcha completion, clean up channel messages after 3 min (configurable)

**Comment by hhio618** [comment]
@gaieges Thanks for the clarification. I'll start to work on these adjustments and will add another commit.

File: app/lib/cleanup_worker.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    def run(self):
+        while True:
+            schedule.run_pending()
+            sleep(1)
+
+    @app_context
+    def _cleanup(self):
+        # If no folowup of captcha completion, then we can just cleanup all of messages that are older than self.minutes ago.
+        messages = db.session.query(Message).filter(
+                   Message.created_date<=(datetime.now()-timedelta(minutes=0))
```
Comment (gaieges) [inline app/lib/cleanup_worker.py:37]: `minutes=0`?

File: app/lib/handlers/base.py [hunk 2]
```diff
@@ -100,3 +106,18 @@ def can_restrict_channel(self, bot, chat_id):
...
+        """ Add a message(chat_id, message_id) to the db, so we can delete the message later """
+        db.session.add(
+            Message(
+                chat_id=chat_id,
+                message_id=message_id,
+                user_id=str(user_id),
+            )
+        )
+        db.session.commit()
+        self.logger.info(
```
Comment (gaieges) [inline app/lib/handlers/base.py:120]: This is more of a debug log I'd think

File: app/lib/utils.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
+import logging
+from app.extensions import db
+
+from app.models import Message
+from telegram.error import BadRequest
+
+
+logging.basicConfig(
```
Comment (gaieges) [inline app/lib/utils.py:8]: Not sure if we have a centralized logger for this but typically you only do this once per app, not in each file (unless its a cli or standalone script)

File: app/models.py [hunk 4]
```diff
@@ -39,3 +40,23 @@ def __init__(
...
+    message_id = db.Column(db.String(), nullable=False)
+    created_date = db.Column(db.DateTime, default=datetime.now)
+
+    def __init__(self, user_id, chat_id, message_id):
+        self.user_id = user_id
+        self.chat_id = chat_id
+        self.message_id = message_id
+
+    def __repr__(self):
+        return "<id {}>".format(self.id)
```
Comment (gaieges) [inline app/models.py:62]: Probably better to make this `<Message {id}>`

File: app/tests/bot_handlers/test_handle_start.py [hunk 5]
```diff
@@ -58,8 +61,11 @@ def test(self):
                 ),
             ]
         )
+        # command.add_message_info.assert_called_with("1234", "1234")
+        # self.assertEqual(command.add_message_info.call_count, 2)
```
Comment (gaieges) [inline app/tests/bot_handlers/test_handle_start.py:65]: Perhaps remove all this commented stuff unless you think it'd be useful for the future

**Review by gaieges — COMMENTED** [review]
This whole repo seems to need a lint tool added to it.

Aside from that this looks pretty good.  This should be able to be rolled back no problem so we can deploy, test it out and rollback if needed.

Is there a beta / test deployment we have here @posix4e ?

File: app/lib/cleanup_worker.py (hunk 1 above)
Comment (hhio618) [reply app/lib/cleanup_worker.py:37]: Sorry that's my mistake while testing

File: app/tests/bot_handlers/test_handle_start.py (hunk 5 above)
Comment (hhio618) [reply app/tests/bot_handlers/test_handle_start.py:65]: Yeah, it's ok we can remove them for now!

File: app/lib/cleanup_worker.py (hunk 1 above)
Comment (gaieges) [reply app/lib/cleanup_worker.py]: Looks better now, that would have bypassed a lot of the testing!  MAkes me think we might also want a test to make sure messages *arent* deleted  in < 2 min

**Review by gaieges — APPROVED** [review]
Lgtm, need to figure out beta testing / deployment testing, deferring to @posix4e for that

File: Pipfile [hunk 6]
```diff
@@ -11,14 +11,15 @@ coverage = "*"
 [packages]
 flask = ">=1.1"
 python-telegram-bot = "*"
-request = "*"
+requests = "*"
```
Comment (posix4e) [inline Pipfile:14]: Oh dear

**Comment by gaieges** [comment]
@hhio618 can you rebase your changes from master? I enabled tests in this repo so this will help ensure they run before merging this.

**Comment by hhio618** [comment]
@gaieges done

**Comment by uivlis** [comment]
@gaieges can I merge and payout this?

**Comment by gaieges** [comment]
@hhio618 awesome! Looks like your tests pass and look good. 

Can you spin up a test bot so we can validate it works as intended?  If so we can close this out and get you paid.

**Comment by hhio618** [comment]
Thanks, @gaieges. No problem, I can run the bot localy in polling mode and that will be enough to test the periodic cleanup functionality. 
For testing "successful siteverify" functionality, we need some sort of hosting environment.

**Comment by gaieges** [comment]
> For testing "successful siteverify" functionality, we need some sort of hosting environment.

True can you set it up in a heroku instance and give it a proper hcaptcha sitekey? There are some details on how to do that in the readme

**Comment by hhio618** [comment]
Yeah, here is the bot @hhio618_bot ready for testing!

**Comment by gaieges** [comment]
`hhio618_hcaptcha` I think you mean right?   Seems it's not responding to me, does it work for you?

**Comment by hhio618** [comment]
> `hhio618_hcaptcha` I think you mean right? Seems it's not responding to me, does it work for you?
Yes, I meant hhio618_hcaptcha. I restarted the bot instance on heroku. I think now it is ready for test.

**Comment by hhio618** [comment]
> > > `hhio618_hcaptcha` I think you mean right? Seems it's not responding to me, does it work for you?
> > 
> > 
> > Yes, I meant hhio618_hcaptcha. I restarted the bot instance on Heroku. I think now it is ready for the test.
> 
> Does it work for you? Still not responding for me

I think maybe there is a problem with Heroku or bot: "Error R12 (Exit timeout) -> At least one process failed to exit within 30 seconds of SIGTERM"[
[log.txt](https://github.com/hCaptcha/telegram-bot/files/5394363/log.txt)
](url)

**Comment by gaieges** [comment]
Oh finally got some responses from it, all at once.  I'm sure related to above. 

Was able to test out the captcha and finish the process, and see the cleaned up messages.  I think we're good here.  Thanks for all the help!

**Comment by hhio618** [comment]
Glad to help.

**Pull request merged**